feat: nextest profile inheritance

asder8215 opened the pull request:
This PR implements a way for nextest profiles to inherit from other profiles as mentioned in #387.

asder8215 commented:
Isn't it possible for there to be a self referential SCC if the user sets the inherits to the current profile?

sunshowers commented:
Ah I think we should just reject that case separately, no good reason to even try to model it or support it.

sunshowers commented:
I think this can be *way* simpler than what you're imagining. There are 4 properties which need to be verified:

1. The default profile cannot inherit from any other profile. (And in fact, any profiles under default- cannot inherit from any other profile.)
2. A profile cannot inherit from itself.
3. A profile cannot inherit from a profile that isn't defined.

1, 2, and 3 don't need the construction of a graph at all, so check for them at the beginning. Then, construct a _reduced_ graph with each node being a profile and each edge being an inherits relationship not rejected by any of the 1/2/3 checks above. (Include the implicit inherits relationship from each non-default profile to the default profile if `inherits` is not defined.) You can probably combine all 3 properties + graph construction into the same loop.

Then, the only remaining check is:

4. In this reduced graph, there are no SCCs with 2 or more nodes.

These 4 properties together should establish the overall invariant that the inheritance graph is well-formed (acyclic, and nothing points to unknown nodes.) You don't need to remove nodes or anything at any point, just break it up into graph and non-graph checks.

To report errors here I'd use an "error collector" or "diagnostic sink" pattern. Pass in a `&mut Vec<InheritError>` or something where `InheritError` is an enum with variants for all the properties, then push to that vector on seeing an error.

Hope this helps.

asder8215 commented:
@sunshowers, I modified the code to follow the properties we talked about earlier and added a couple of test cases. Let me know if there's anything I should change or if there's any other test case I should add here!

sunshowers commented on nextest-runner/src/config/core/imp.rs:
```diff
@@ -1154,6 +1176,151 @@ impl NextestConfigImpl {
+                inherit_err_collector.push(InheritsError::DefaultProfileInheritance(
+                    default_profile.to_string(),
+                ));
+            }
```
Don't quite follow what's happening here -- we should prevent any profiles whose name starts with `default-` from specifying `inherits`.

asder8215 replied:
Sounds good. I misunderstood the no inheritance on default profiles as regarding only the default profile names reserved in `DEFAULT_PROFILES` array.

I'll just do a sanity check on the `DefaultProfileImpl` here (which I believe works on the `default` profile) in this function, and then handle all the other default profiles  that starts with `default-` within `check_inheritance_cycles()`.

sunshowers commented on nextest-runner/src/config/elements/inherits.rs:
```diff
@@ -0,0 +1,238 @@
+        inherits: Option<String>,
+    }
+
+    #[test_case(
```
good tests!

asder8215 commented on nextest-runner/src/config/elements/inherits.rs:
```diff
@@ -0,0 +1,238 @@
+    }
+}
+
+// TODO: Need to write test cases  for this
```
Forgot to remove this from a previous commit

asder8215 commented:
It's all good and no pressure or rush in checking this out.

sunshowers commented on nextest-runner/src/config/elements/inherits.rs:
```diff
@@ -0,0 +1,238 @@
+        junit: JunitImpl,
+        archive: Option<ArchiveConfig>,
+        inherits: Option<String>,
+    }
```
Ended up shortening this to just the fields being tested in the test below.
